src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs

```diff
@@ -28,56 +27,80 @@ internal static partial class ConflictResolver
                 ImmutableArray<SymbolKey> nonConflictSymbolKeys,
+                ImmutableDictionary<DocumentId, DocumentRenameInfo> documentIdToRenameInfo,
+                ImmutableDictionary<ISymbol, string> symbolToReplacementText,
+                ImmutableDictionary<ISymbol, bool> symbolToReplacementTextValid,
+                ImmutableDictionary<ISymbol, (DocumentId declarationDocumentId, Location declarationLocation)> symbolToDeclarationDocumentAndLocation,
+                CodeCleanupOptionsProvider fallBackOptions,
                 CancellationToken cancellationToken)
             {
-                _renameLocationSet = renameLocationSet;
-                _renameSymbolDeclarationLocation = renameSymbolDeclarationLocation;
-                _originalText = renameLocationSet.Symbol.Name;
-                _replacementText = replacementText;
-                _nonConflictSymbolKeys = nonConflictSymbolKeys;
+                var dependencyGraph = solution.GetProjectDependencyGraph();
+                _symbolToRenameLocations = symbolicRenameLocations.ToImmutableDictionary(symbolicRenameLocations => symbolicRenameLocations.Symbol);
+                _baseSolution = solution;
+                _topologicallySortedProjects = dependencyGraph.GetTopologicallySortedProjects(cancellationToken).ToImmutableArray();
                 _cancellationToken = cancellationToken;
+                _nonConflictSymbolKeys = nonConflictSymbolKeys;
+                _fallbackOptions = fallBackOptions;
 
-                _conflictLocations = SpecializedCollections.EmptySet<ConflictLocationInfo>();
-                _replacementTextValid = true;
-
-                // only process documents which possibly contain the identifiers.
-                _documentIdOfRenameSymbolDeclaration = renameLocationSet.Solution.GetRequiredDocument(renameSymbolDeclarationLocation.SourceTree!).Id;
-
-                _renameAnnotations = new AnnotationTable<RenameAnnotation>(RenameAnnotation.Kind);
+                _documentIdToRenameInfo = documentIdToRenameInfo;
+                _documentsIdsToBeCheckedForConflict = documentIdToRenameInfo.Keys.ToImmutableHashSet();
+                _symbolToReplacementText = symbolToReplacementText;
+                _symbolToReplacementTextValid = symbolToReplacementTextValid;
+                _symbolToDeclarationDocumentAndLocation = symbolToDeclarationDocumentAndLocation;
             }
 
-            private SymbolRenameOptions RenameOptions => _renameLocationSet.Options;
-            private CodeCleanupOptionsProvider FallbackOptions => _renameLocationSet.FallbackOptions;
+            /// <summary>
+            /// Whether the <param name="newReferencedSymbol"/> has conflict with <param name="renameDeclarationLocationReference"/>
+            /// </summary>
+            protected abstract bool HasConflictForMetadataReference(RenameDeclarationLocationReference renameDeclarationLocationReference, ISymbol newReferencedSymbol);
```

Comment:
The only thing based on the 'Single Symbol' rename assumption.
I extract the implementation logic to 'SingleSymbolRenameSession'.
And to rename 'multiple symbols', we just need to create a derived type from 'Session', and implement this function.